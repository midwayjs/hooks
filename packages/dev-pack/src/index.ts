export * from './middleware'
export * from './server'
export * from './share'
