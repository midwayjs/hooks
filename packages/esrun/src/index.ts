import { readDefaultTsConfig } from '@swc-node/register/read-default-tsconfig'
import { register as registerSWC } from '@swc-node/register/register'

export function register() {
  const options = readDefaultTsConfig()

  // ensure all files are compiled
  options.files = null
  registerSWC(options)

  if (options.paths && Object.keys(options.paths).length) {
    require('tsconfig-paths/register')
  }
}
