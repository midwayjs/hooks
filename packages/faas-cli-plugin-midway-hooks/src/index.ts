export { MidwayHooksPlugin } from './plugin'
export { ViewRoutesPlugin } from './routes'
// export { InstallPlugin } from './installer'
