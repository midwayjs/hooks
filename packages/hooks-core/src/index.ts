export * from './runtime'
export * from './router'
export * from './types'
export * from './common'

export * from './api'
export * from './adapter'

export { default as urlJoin } from 'proper-url-join'
