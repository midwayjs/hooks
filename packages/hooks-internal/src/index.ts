export * from './config'
export * from './root'
export * from './router'
export * from './router/file'
export * from './util'
export * from './hydrate'
export * from './const'
