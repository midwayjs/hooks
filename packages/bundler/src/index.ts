import { createUnplugin, UnpluginOptions } from 'unplugin'
import {
  AbstractRouter,
  ApiRoute,
  createDebug,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  parseApiModule,
} from '@midwayjs/hooks-core'
import '@midwayjs/esrun/register'

export { EsbuildPlugin, VitePlugin, RollupPlugin } from 'unplugin'

const debug = createDebug('hooks-bundler')

const omit = (key: string, { [key]: _, ...obj }) => obj

interface BundlerConfig {
  name: string
  router: AbstractRouter
}

export abstract class AbstractBundlerAdapter {
  constructor(private config: BundlerConfig) {}

  getName() {
    return this.config.name
  }

  abstract getSource(): string

  getRouter() {
    return this.config.router
  }

  getUnplugin(): Partial<UnpluginOptions> {
    return {}
  }

  transformApiRoutes(apis: ApiRoute[]) {
    return apis
  }

  generateClient(apis: ApiRoute[]) {
    let fetcherId = 0
    const importClient = new Map<string, [string, string, number]>()

    const importCodes = new Set<string>()
    const functionCodes = []

    for (const {
      trigger,
      functionName,
      functionId,
      useInputMetadata,
    } of apis) {
      if (!trigger.requestClient) continue
      const { fetcher, client } = trigger.requestClient

      const clientId = `${client}_${fetcher}`
      if (!importClient.has(clientId)) {
        importClient.set(clientId, [client, fetcher, fetcherId++])
      }

      const [clientName, fetcherName, idx] = importClient.get(clientId)
      const uniqueFetcher = `rpc$${idx}`
      importCodes.add(
        `import { ${fetcherName} as ${uniqueFetcher} } from '${clientName}'`
      )

      const exportModifier =
        functionName === EXPORT_DEFAULT_FUNCTION_ALIAS ? 'default' : ''
      functionCodes.push(`
        export ${exportModifier} function ${functionName} (...args) {
          const route = ${functionName}.route;
          return ${uniqueFetcher}(...args, route)
        };
        ${functionName}.route = {
          trigger: ${JSON.stringify(omit('requestClient', trigger))},
          functionId: '${functionId}',
          useInputMetadata: ${!!useInputMetadata},
        };
      `)
    }

    const code = [
      '// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.',
      ...importCodes,
      ...functionCodes,
    ].join('\n')

    debug('generate client code: %s', code)

    return code
  }
}

export function createBundlerPlugin(adapter: AbstractBundlerAdapter) {
  return createUnplugin((options) => {
    return {
      name: adapter.getName(),
      enforce: 'pre',
      transformInclude(id: string) {
        const source = adapter.getSource()
        const router = adapter.getRouter()

        if (!router.isSourceFile(id, source)) {
          return false
        }

        return router.isApiFile({
          file: id,
          mod: requireWithoutCache(id),
        })
      },
      async transform(code: string, id: string) {
        const apis = parseApiModule(
          requireWithoutCache(id),
          id,
          adapter.getRouter()
        )
        const transformedApis = adapter.transformApiRoutes(apis)
        return adapter.generateClient(transformedApis)
      },
      ...adapter.getUnplugin(),
    }
  })
}

/**
 * In dev mode, ensure get latest module content when file changes
 */
export function requireWithoutCache(module: string) {
  delete require.cache[require.resolve(module)]
  return require(module)
}
