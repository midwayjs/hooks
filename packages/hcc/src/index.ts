import { getRouter, getSource, setProjectRoot } from '@midwayjs/hooks-internal'
import { GenerateTarget, getEntryCode } from './midway'

export async function getPreloadCode(cwd: string) {
  setProjectRoot(cwd)

  const source = getSource({ useSourceFile: false })
  const router = getRouter(source)

  return `
    ${getEntryCode(source, router, GenerateTarget.JS)}
    module.exports = require('./configuration');
  `
}
