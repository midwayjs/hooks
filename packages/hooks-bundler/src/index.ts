import { createBundlerPlugin } from '@midwayjs/bundler'
import { MidwayBundlerAdapter } from './adapter'

const plugin = createBundlerPlugin(new MidwayBundlerAdapter())

export const webpack = plugin.webpack
export const vite = plugin.vite
export const esbuild = plugin.esbuild
export const rollup = plugin.rollup

export const ServerlessBundlerPlugin = {
  get vite() {
    throw new Error(
      `ServerlessBundlerPlugin is deprecated, use import { vite } from '@midwayjs/hooks-bundler' directly.`
    )
  },
}
