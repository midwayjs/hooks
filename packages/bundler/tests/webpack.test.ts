import { ApiRouter } from '@midwayjs/hooks-core'
import { wrap } from 'jest-snapshot-serializer-raw'
import Memoryfs from 'memory-fs'
import { resolve as pathResolve } from 'path'
import webpack from 'webpack'
import { AbstractBundlerAdapter, createBundlerPlugin } from '../src'
import prettier from 'prettier'

const formatCode = (code: string) => {
  return prettier.format(code, {
    semi: true,
    singleQuote: true,
    parser: 'babel',
  })
}

const root = pathResolve(__dirname, './fixtures/base-app')

class TestBundlerAdapter extends AbstractBundlerAdapter {
  getSource(): string {
    return root
  }
}

const router = new ApiRouter()

const testBundlerAdapter = new TestBundlerAdapter({
  name: 'test',
  router,
})

const { webpack: WebpackPlugin } = createBundlerPlugin(testBundlerAdapter)

describe('unplugin-hooks webpack', () => {
  const getEntry = (path: string) => {
    return pathResolve(root, path)
  }

  test('Compile render', async () => {
    const output = await compile(root, getEntry('render/index.ts'))
    expect(wrap(formatCode(output))).toMatchSnapshot()
  })

  test('Compile lambda', async () => {
    const output = await compile(root, getEntry('lambda/index.ts'))
    expect(wrap(formatCode(output))).toMatchSnapshot()
  })

  test('the second build should match the first.', async () => {
    const first = await compile(root, getEntry('lambda/index.ts'))
    const second = await compile(root, getEntry('lambda/index.ts'))
    expect(first).toEqual(second)
  })

  test('non-lambda files should not be compiled', async () => {
    const output = await compile(root, getEntry('util/util.ts'))
    expect(wrap(formatCode(output))).toMatchSnapshot()
  })
})

function compile(context: string, entry: string): Promise<string> {
  return new Promise((resolve) => {
    const compiler = webpack({
      context,
      entry,
      plugins: [WebpackPlugin()],
      output: {
        path: '/',
        filename: 'client.js',
      },
      externals: {
        '@midwayjs/rpc': 'fetch',
        axios: 'axios',
      },
      mode: 'development',
      module: {
        rules: [
          {
            test: /\.tsx?$/,
            use: [
              {
                loader: 'ts-loader',
                options: {
                  transpileOnly: true,
                },
              },
            ],
          },
        ],
      },
    })

    compiler.outputFileSystem = new Memoryfs()

    compiler.run((err: any, stats) => {
      if (err) {
        console.error(err.stack || err)
        if (err.details) {
          console.error(err.details)
        }
        return
      }

      const info = stats.toJson({ source: true })

      if (stats.hasErrors()) {
        console.error(info.errors)
      }

      if (stats.hasWarnings()) {
        console.warn(info.warnings)
      }

      const module = info.modules.find((mod) => mod.identifier.includes(entry))
      resolve(module.source.toString())
    })
  })
}
