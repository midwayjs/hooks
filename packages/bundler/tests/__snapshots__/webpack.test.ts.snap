// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unplugin-hooks webpack Compile lambda 1`] = `
// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.
import { http as rpc$0 } from '@midwayjs/rpc';
export default function $default(...args) {
  const route = $default.route;
  return rpc$0(...args, route);
}
$default.route = {
  trigger: { type: 'HTTP', method: 'GET', path: '/api/index' },
  functionId: 'index',
  useInputMetadata: false,
};
export function usePost(...args) {
  const route = usePost.route;
  return rpc$0(...args, route);
}
usePost.route = {
  trigger: { type: 'HTTP', method: 'POST', path: '/api/usePost' },
  functionId: 'usePost',
  useInputMetadata: false,
};
export function withQuery(...args) {
  const route = withQuery.route;
  return rpc$0(...args, route);
}
withQuery.route = {
  trigger: { type: 'HTTP', method: 'GET', path: '/api/withQuery' },
  functionId: 'withQuery',
  useInputMetadata: true,
};

`;

exports[`unplugin-hooks webpack Compile render 1`] = `
// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.
import { http as rpc$0 } from '@midwayjs/rpc';
export function bar(...args) {
  const route = bar.route;
  return rpc$0(...args, route);
}
bar.route = {
  trigger: { type: 'HTTP', method: 'GET', path: '/api/bar' },
  functionId: 'bar',
  useInputMetadata: false,
};
export default function $default(...args) {
  const route = $default.route;
  return rpc$0(...args, route);
}
$default.route = {
  trigger: { type: 'HTTP', method: 'GET', path: '/*' },
  functionId: 'index',
  useInputMetadata: false,
};
export function foo(...args) {
  const route = foo.route;
  return rpc$0(...args, route);
}
foo.route = {
  trigger: { type: 'HTTP', method: 'GET', path: '/api/foo' },
  functionId: 'foo',
  useInputMetadata: false,
};

`;

exports[`unplugin-hooks webpack non-lambda files should not be compiled 1`] = `
export function isTrue(value) {
  return value === 'true' || value === true;
}

`;
